src/utility/api.js

```diff
@@ -0,0 +1,153 @@
+/**
+ * Question: why does this file exist?
+ * Answer: The fetch API, while great, has limitations. In particular, it does not throw a rejected promise when
+ * a non 200 response is received and does not automatically unmarshal JSON. This wrapper solves those problems and
+ * provides a uniform API, returning an object of the form { json, error, response } in all cases, whether the promise
+ * was rejected or not. In some cases, we still want access to the json even if the response is not 200.
+ * 
+ * This gives you the original response as well, in case you want that.
+ * 
+ * This wrapper is also designed to interface nicely with the URL API implemented in url.js.
+ * 
+ * This wrapper will *not* work as intended for response types that are not JSON or empty.
+ */
+
+const REQUEST_VERBS = Object.freeze({
+    GET: "GET",
+    PUT: "PUT",
+    POST: "POST",
+    DELETE: "DELETE",
+})
+
+const NOT_JSON_AND_OK = "NOT_JSON_AND_OK";
+const NOT_JSON_AND_NOT_OK = "NOT_JSON_AND_NOT_OK";
+const JSON_AND_NOT_OK = "JSON_AND_NOT_OK";
+
+function makeResponseData(response, json, error) {
```

Comment:
Preliminary justification for such a function: helps to enforce the ensured interface for what the functions return. Everyone hates `new`.

Reply:
sgtm